### PR TITLE
feat: configurable NATS connection retries

### LINKED ISSUE #87: No support for NATS connection retries
It would be useful to support configuration options for NATS connection retries in the event that a marmot follower is initialized slightly before its leader. In our use case, two hosts are provisioned simultaneously and the follower host occasionally lags the leader by as much as a few minutes.

Marmot follower error when attempting to connect prior to leader initialization:
```shell
Oct 17 00:40:04 two-node-two marmot[1693]: 12:40AM DBG Opening database node_id=2197861447266130575 path=/var/lib/rancher/k3s/server/db/state.db
Oct 17 00:40:04 two-node-two marmot[1693]: 12:40AM DBG Forcing WAL checkpoint node_id=2197861447266130575
Oct 17 00:40:07 two-node-two marmot[1693]: 12:40AM PNC Unable to initialize snapshot storage error="dial tcp X.X.X.X:4222: connect: no route to host" node_id=2197861447266130575
Oct 17 00:40:07 two-node-two marmot[1693]: panic: Unable to initialize snapshot storage
Oct 17 00:40:07 two-node-two marmot[1693]: goroutine 1 [running]:
Oct 17 00:40:07 two-node-two marmot[1693]: github.com/rs/zerolog/log.Panic.(*Logger).Panic.func1({0x1158d4c?, 0x0?})
Oct 17 00:40:07 two-node-two marmot[1693]:         /home/runner/go/pkg/mod/github.com/rs/zerolog@v1.29.1/log.go:376 +0x27
Oct 17 00:40:07 two-node-two marmot[1693]: github.com/rs/zerolog.(*Event).msg(0xc000282300, {0x1158d4c, 0x25})
Oct 17 00:40:07 two-node-two marmot[1693]:         /home/runner/go/pkg/mod/github.com/rs/zerolog@v1.29.1/event.go:156 +0x2c2
Oct 17 00:40:07 two-node-two marmot[1693]: github.com/rs/zerolog.(*Event).Msg(...)
Oct 17 00:40:07 two-node-two marmot[1693]:         /home/runner/go/pkg/mod/github.com/rs/zerolog@v1.29.1/event.go:108
Oct 17 00:40:07 two-node-two marmot[1693]: main.main()
Oct 17 00:40:07 two-node-two marmot[1693]:         /home/runner/work/marmot/marmot/marmot.go:66 +0x70a
Oct 17 00:40:07 two-node-two systemd[1]: marmot.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Oct 17 00:40:07 two-node-two systemd[1]: marmot.service: Failed with result 'exit-code'.
Oct 17 00:40:07 two-node-two systemd[1]: marmot.service: Scheduled restart job, restart counter is at 2.
Oct 17 00:40:07 two-node-two systemd[1]: Stopped Marmot synchronizes the k8s state in SQLite between nodes in a two node topology.
Oct 17 00:40:07 two-node-two systemd[1]: Started Marmot synchronizes the k8s state in SQLite between nodes in a two node topology.
Oct 17 00:40:07 two-node-two marmot[1699]: 12:40AM DBG Opening database node_id=2197861447266130575 path=/var/lib/rancher/k3s/server/db/state.db
Oct 17 00:40:07 two-node-two marmot[1699]: 12:40AM DBG Forcing WAL checkpoint node_id=2197861447266130575
Oct 17 00:40:10 two-node-two marmot[1699]: 12:40AM PNC Unable to initialize snapshot storage error="dial tcp X.X.X.X:4222: connect: no route to host" node_id=2197861447266130575
```

### PR DESCRIPTION
**Pull request by TylerGillson**
Add configuration options & logic for NATS connection retries. Fixes #87.

In our use case, the NATS worker may boot before the NATs leader on occasion, hence these options would be very useful!

### CODE REVIEW

**Comment by maxpert** [comment]
Thanks for the build, but I anticipate that there should be a way to configure it in NATS configuration file and people can do it via that rather than adding additional bloat in configuration. Can you check if that is possible via nats configuration file?

**Comment by TylerGillson** [comment]
> Thanks for the build, but I anticipate that there should be a way to configure it in NATS configuration file and people can do it via that rather than adding additional bloat in configuration. Can you check if that is possible via nats configuration file?

@maxpert My interpretation of the NATS client library is that the [reconnection options](https://github.com/nats-io/nats.go/blob/main/nats.go#L329) are only applicable _after_ an initial connection is established.

And I don't see anything promising in the [server options](https://github.com/nats-io/nats-server/blob/main/server/opts.go#L249) either. I will nevertheless toy around with the reconnection options and let you know.

**Comment by TylerGillson** [comment]
@maxpert your anticipation was correct 😁

However, I still don't see a client configuration file or parser for same in the NATS library.

What do you think about the changes now? Only one new configuration option.

**Comment by maxpert** [comment]
You might be right. Let me explore myself too.

**Comment by TylerGillson** [comment]
> You might be right. Let me explore myself too.

Ok - I did validate this code though, FWIW. My earlier comment was uninformed. The client retries work fine on the initial connection with these changes.

File: cfg/config.go [hunk 1]
```diff
@@ -67,17 +67,18 @@ type SnapshotConfiguration struct {
...
+	StreamPrefix            string   `toml:"stream_prefix"`
+	ServerConfigFile        string   `toml:"server_config"`
+	SeedFile                string   `toml:"seed_file"`
+	CredsUser               string   `toml:"user_name"`
+	CredsPassword           string   `toml:"user_password"`
+	CAFile                  string   `toml:"ca_file"`
+	CertFile                string   `toml:"cert_file"`
+	KeyFile                 string   `toml:"key_file"`
+	BindAddress             string   `toml:"bind_address"`
+	ConnectTotalWaitMinutes int      `toml:"connect_total_wait_minutes"`
```
Comment (maxpert) [inline cfg/config.go:81]: I will have it consistent with other time units I believe everywhere it's seconds so this should be seconds too. We should rename it to something like ConnectionTimeout I believe

File: stream/nats.go [hunk 2]
```diff
@@ -76,3 +75,31 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
...
+		nats.Timeout(10 * time.Second),
+		nats.ClosedHandler(func(nc *nats.Conn) {
+			log.Fatal().Msg(fmt.Sprintf("Exiting: %v", nc.LastError()))
+		}),
+	}
+	if cfg.Config.NATS.ConnectTotalWaitMinutes == 0 {
+		return opts
+	}
+
+	totalWait := time.Duration(cfg.Config.NATS.ConnectTotalWaitMinutes) * time.Minute
```
Comment (maxpert) [inline stream/nats.go:91]: This will change with what I mentioned above

File: stream/nats.go [hunk 3]
```diff
@@ -76,3 +75,31 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
...
+
+func setupConnOptions() []nats.Option {
+	opts := []nats.Option{
+		nats.Name(cfg.Config.NodeName()),
+		nats.Timeout(10 * time.Second),
+		nats.ClosedHandler(func(nc *nats.Conn) {
+			log.Fatal().Msg(fmt.Sprintf("Exiting: %v", nc.LastError()))
+		}),
+	}
+	if cfg.Config.NATS.ConnectTotalWaitMinutes == 0 {
```
Comment (maxpert) [inline stream/nats.go:87]: Ditto as top

**Review by maxpert — CHANGES_REQUESTED** [review]
Change time units, I will have to give it another pass once done. I think this PR makes sens.e

File: cfg/config.go (hunk 1 above)
Comment (TylerGillson) [reply cfg/config.go]: I've updated to `ConnectRetries` as I thought that made more sense. Happy to rename it again as you see fit.

File: cfg/config.go [hunk 4]
```diff
@@ -67,18 +67,18 @@ type SnapshotConfiguration struct {
...
-	StreamPrefix            string   `toml:"stream_prefix"`
-	ServerConfigFile        string   `toml:"server_config"`
-	SeedFile                string   `toml:"seed_file"`
-	CredsUser               string   `toml:"user_name"`
-	CredsPassword           string   `toml:"user_password"`
-	CAFile                  string   `toml:"ca_file"`
-	CertFile                string   `toml:"cert_file"`
-	KeyFile                 string   `toml:"key_file"`
-	BindAddress             string   `toml:"bind_address"`
-	ConnectTotalWaitMinutes int      `toml:"connect_total_wait_minutes"`
```
Comment (maxpert) [inline cfg/config.go]: You need to assign a default value to this one, may be a default value of 5 will be good enought. Checkout where I initialize BindAddress and other values.

File: stream/nats.go [hunk 5]
```diff
@@ -77,29 +77,22 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
 }
 
 func setupConnOptions() []nats.Option {
-	opts := []nats.Option{
+	// total wait = ( default nats.Timeout (2s) + nats.ReconnectWait (1s) ) * connect_retries
```
Comment (maxpert) [inline stream/nats.go:80]: Yeah after this comment I am convinced we should allow configuring timeout too.

File: stream/nats.go [hunk 6]
```diff
@@ -77,29 +77,22 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
...
 		nats.Name(cfg.Config.NodeName()),
-		nats.Timeout(10 * time.Second),
 		nats.ClosedHandler(func(nc *nats.Conn) {
 			log.Fatal().Msg(fmt.Sprintf("Exiting: %v", nc.LastError()))
 		}),
+		nats.RetryOnFailedConnect(true),
+		nats.ReconnectWait(time.Second),
+		nats.MaxReconnects(cfg.Config.NATS.ConnectRetries),
+		nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
+			log.Printf("Disconnected due to: %v, will attempt to reconnect for %d seconds", err, totalWait)
```
Comment (maxpert) [inline stream/nats.go:92]: Use: `log.Error().Msg("...)`

File: stream/nats.go [hunk 7]
```diff
@@ -77,29 +77,22 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
...
+		}),
+		nats.ReconnectHandler(func(nc *nats.Conn) {
+			log.Printf("Reconnected to [%s]", nc.ConnectedUrl())
```
Comment (maxpert) [inline stream/nats.go:95]: Use: `log.Info().Msg("...)`

**Review by maxpert — CHANGES_REQUESTED** [review]
Couple of more things to fine tune.

File: stream/nats.go (hunk 6 above)
Comment (maxpert) [reply stream/nats.go:92]: use Err and Int to write structured logs on `err` and `totalWait`

**Comment by TylerGillson** [comment]
@maxpert apologies for the mixup but my initial take on the NATS options was actually correct. The reconnection options seem to only be applicable after an initial connection is established.

I was testing with the wrong binary 🤦🏼‍♂️

So I'll be reverting this PR back to the original approach. I'd appreciate any insight or opinions you have from your own testing.

**Comment by TylerGillson** [comment]
@maxpert - it is working as expected now:
image

**Comment by maxpert** [comment]
Perfect I will run couple of tests and try it in my sandbox this coming week.

File: stream/nats.go [hunk 8]
```diff
@@ -35,10 +33,26 @@ func Connect() (*nats.Conn, error) {
...
+		}
+
+		log.Warn().
+			Err(err).
+			Int("attempt", i+1).
+			Int("attempt_limit", cfg.Config.NATS.ConnectRetries).
+			Str("status", conn.Status().String()).
+			Msg("NATS connection failed")
+
+		continue
```
Comment (maxpert) [inline stream/nats.go:52]: Don't need this continue

File: stream/nats.go [hunk 9]
```diff
@@ -76,3 +90,27 @@ func getNatsTLSFromConfig() ([]nats.Option, error) {
 
 	return opts, nil
 }
+
+func setupConnOptions() []nats.Option {
+	return []nats.Option{
+		nats.Name(cfg.Config.NodeName()),
+		nats.RetryOnFailedConnect(true),
+		nats.ReconnectWait(time.Second),
```
Comment (maxpert) [inline stream/nats.go:98]: Let's add option in configuration to customize this as well. So that people can configure wait time.

**Review by maxpert — COMMENTED** [review]
Tested and few minor nits I want to be ironed out

**Comment by TylerGillson** [comment]
@maxpert I added `ReconnectWaitSeconds` per your request, but haven't exposed timeout because I was trying to avoid configuration bloat. Currently the default timeout of 2s is set for the dialer for each connection attempt.

Would you like me to expose that option too?

**Review by maxpert — APPROVED** [review]

**Pull request merged**